ipm_node:
  ros__parameters:

    enable_vis: true
    
    # ----------------------------------------------------
    # 1. 话题设置 (Topics)
    # ----------------------------------------------------
    # 订阅的原始/畸变校正后的图像话题
    image_topic: "/camera/color/image_raw" 
    
    # 发布的鸟瞰图障碍物点云话题 (给 Costmap)
    pointcloud_topic: "/bev/obstacles"
    
    # (可选) 发布的调试用鸟瞰图图像话题
    bev_debug_image_topic: "/bev/debug_image"

    # === 相机标定 ===
    camera_matrix: [560.08, 0.0, 313.60, 
                    0.0, 560.16, 226.29, 
                    0.0, 0.0, 1.0]
    distortion_coeffs: [0.1248, -0.1525, -0.0104, -0.0031, 0.0]


    # ----------------------------------------------------
    # 2. 鸟瞰图 (BEV) 尺寸
    # ----------------------------------------------------
    # 输出的鸟瞰图的像素尺寸 (高和宽)
    bev_width: 640
    bev_height: 480

    # ######最初小矩形
    # src_points: [201.0, 357.0, 442.0, 356.0, 558.0, 447.0, 89.0, 452.0]
    # dst_points: [0.0, 0.0, 640.0, 0.0, 640.0, 480.0, 0.0, 480.0]
    # world_width_m: 0.244
    # world_height_m: 0.273
    # origin_offset_x_m: 0.325  # BEV "最近边" 在车前多远
    # origin_offset_y_m: 0.122  # BEV "左侧边" 在车左多远
    # ######大矩形
    # src_points: [180.0, 305.0, 453.0, 307.0, 632.0, 389.0, 23.0, 395.0]
    # dst_points: [0.0, 0.0, 640.0, 0.0, 640.0, 480.0, 0.0, 480.0]    
    # world_width_m: 0.436
    # world_height_m: 0.5095
    # origin_offset_x_m: 0.325  # BEV "最近边" 在车前多远
    # origin_offset_y_m: 0.205 # BEV "左侧边" 在车左多远 0.218- 0.013

    ######新避障矩形
    src_points: [155.0, 311.0, 464.0, 313.0, 579.0, 358.0, 47.0, 360.0]
    dst_points: [0.0, 0.0, 640.0, 0.0, 640.0, 480.0, 0.0, 480.0]
    world_width_m: 0.448
    world_height_m: 0.324
    origin_offset_x_m: 0.457  # BEV "最近边" 在车前多远
    origin_offset_y_m: 0.229  # BEV "左侧边" 在车左多远    


    # # ----------------------------------------------------
    # # 3. 逆透视变换 (IPM) 标定 (!!! 必须自定义 !!!)
    # # ----------------------------------------------------
    # # A. 源点 (Source Points):
    # #    在相机图像 (640x480) 中手动选择的4个 *地面* 点 [u, v]
    # #    顺序: [左上, 右上, 右下, 左下] (例如一个梯形)
    # src_points: [201.0, 357.0, 442.0, 356.0, 558.0, 447.0, 89.0, 452.0]
    # # [(201, 357), (442, 356), (558, 447), (89, 452)]

    # # B. 目标点 (Destination Points):
    # #    这4个点在鸟瞰图中的 *像素* 坐标
    # #    通常映射为一个完美的矩形。
    # #    顺序必须与 src_points 对应: [左上, 右上, 右下, 左下]
    # dst_points: [0.0, 0.0, 640.0, 0.0, 640.0, 480.0, 0.0, 480.0]

    # # ----------------------------------------------------
    # # 4. 物理尺寸标定 (!!! 必须自定义 !!!)
    # # ----------------------------------------------------
    # # 标定矩形的真实物理尺寸 (米)
    # world_width_m: 0.244
    # world_height_m: 0.273

    # # BEV 原点 (即 dst_points 的 [0, 480] 点) 在 base_link 中的位置
    # # (假设 base_link x 轴朝前, y 轴朝左)
    # origin_offset_x_m: 0.325  # BEV "最近边" 在车前多远
    # origin_offset_y_m: 0.122  # BEV "左侧边" 在车左多远

    # ----------------------------------------------------
    # 6. 红色检测 (HSV 阈值) (!!! 需要微调 !!!)
    # ----------------------------------------------------
    # H (色相): 0-179, S (饱和度): 0-255, V (明度): 0-255
    # hsv_lower_red1: [0, 100, 100]
    # hsv_upper_red1: [10, 255, 255]
    # hsv_lower_red2: [160, 100, 100]
    # hsv_upper_red2: [179, 255, 255]


    # hsv = cv2.cvtColor(img_eq, cv2.COLOR_BGR2HSV)
    # mean_v = np.mean(hsv[:, :, 2])
    # v_thresh = max(50, int(mean_v * 0.5))
    # lower_red_1 = np.array([0, 90, v_thresh])
    # upper_red_1 = np.array([10, 255, 255])
    # lower_red_2 = np.array([170, 80, v_thresh])
    # upper_red_2 = np.array([180, 255, 255])
    
    # ----------------------------------------------------
    # 7. 图像处理 (可选调优)
    # ----------------------------------------------------
    # 形态学操作核大小 (用于去噪点和填补孔洞)
    morph_kernel_size: 5